datalad_next/runners/run.py

```diff
@@ -0,0 +1,258 @@
+    protocol_class: type[Protocol],
+    *,
+    cwd: Path | None = None,
+    stdin: int | IO | bytes | Queue[bytes | None] | None = None,
+    timeout: float | None = None,
+    closing_action: Callable | None = None,
+    terminate_time: int | None = None,
+    kill_time: int | None = None,
+    **protocol_kwargs,
+) -> Any | Generator:
+    """ A context manager for subprocesses
+
+    The run-context manager will start a subprocess via ``ThreadedRunner``,
+    provide the result of the subprocess invocation, i.e. the result of
+    ``ThreadedRunner.run()`` in the ``as``-variable, and
+    clean up the subprocess when the context is left.
+
+    The run-context manager supports the guaranteed exit of a subprocess through
+    either:
+
+     a) natural exit of the subprocess
+     b) termination of the subprocess via SIGTERM, if ``terminate_time`` is
+        specified
+     c) termination of the subprocess via SIGKILL, if ``kill_time`` is specified
+
+    If the process terminates the run-context manager will ensure that its exit
+    status is read, in order to prevent zombie processes.
+
+    If neither ``terminate_time`` nor ``kill_time`` are specified, and the
+    subprocess does not exit by itself, for example, because it waits for some
+    input, the ``__exit__``-method of the run-context manager will never return.
+    In other words the thread will seem to "hang" when leaving the run-context.
+    The only way to ensure that the context is eventually left is to provide a
+    ``kill_time``. It is a good idea to provide a ``terminate_time`` in
+    addition, to allow the subprocess a graceful exit (see ``kill_time``- and
+    ``terminate_time``-argument descriptions below).
+
+    Generator- and non-generator-protocols are both supported by the
+    context manager. Depending on the type of the provided protocol the
+    interpretation of ``terminate_time`` and ``kill_time`` are different.
+
+    If a non-generator-protocol is used, counting of the ``terminate_time`` and
+    the ``kill_time`` starts when the subprocess is started.
+
+    If a generator-protocol is used, counting of the ``terminate_time`` and
+    the ``kill_time`` starts when the run-context is left.
+
+    Parameters
+    ----------
+    cmd : list[str]
+        The command list that is passed to ``ThreadedRunner.run()``
+    protocol_class : type[Protocol]
+        The protocol class that should be used to process subprocess events.
+    cwd: Path, optional
+        If provided, defines the current work directory for the subprocess
+    stdin: int | IO | bytes | Queue[bytes | None], optional
+        Input source or data for stdin of the subprocess. See the constructor
+        of :class:`ThreadedRunner` for a detailed description
+    timeout: float, optional
+        If provided, defines the time after which the timeout-callback of the
+        protocol will be enabled. See the constructor
+        of :class:`ThreadedRunner` for a detailed description
```

Comment:
```suggestion
    timeout: float, optional
        If provided, defines the time (in seconds) after which the timeout-callback of the
        protocol will be triggered. See the constructor
        of :class:`ThreadedRunner` for a detailed description
```

It seems that the critical doc is in `Protocol.timeout()`.

> Called if the timeout parameter to WitlessRunner.run() is not `None` and a process file descriptor could not be read (stdout or stderr) or not be written (stdin) within the specified time in seconds, or if waiting for a subprocess to exit takes longer than the specified time.

In general I understand the desire to avoid duplication of documentation. However, the reference `:class:ThreadedRunner` is not resolved, and a readers is tasked with a scavenger hunt of finding that class (which is not in a module of this name, in a different package, which once found points to another class and method which only then has this IMHO critical piece of information).